btas/special/cp_als.h

```diff
@@ -0,0 +1,634 @@
+  typedef typename std::vector<value_type>::const_iterator iterator;
+
+  CP_ALS(Tensor &tensor)
+      : tensor_ref(tensor), dim(tensor_ref.rank()), size(tensor_ref.size()) {
+
+#if not defined(BTAS_HAS_CBLAS) || not defined(_HAS_INTEL_MKL)
+    BTAS_EXCEPTION_MESSAGE(__FILE__, __LINE__,
+                           "CP_ALS requires LAPACKE or mkl_lapack");
+#endif
+#ifdef _HAS_INTEL_MKL
+#include <mkl_trans.h>
+#endif
+    for (int i = 0; i < dim; i++)
+      curr_dim.push_back(tensor_ref.extent(i));
+
+    oldmat.resize(Range{Range1{tensor_ref.extent(0)},
+                        Range1{size / tensor_ref.extent(0)}});
+  }
+
+  ~CP_ALS() = default;
+
+  //Builds rank R approximation starting from r=1 to R
+  //direct method doesn't compute Khatri-Rao intermediate
+  //r_test ||T(0) - U(0)||F where U(1) is the recomposed
+    //approximate tensor and  T(1) is the tensor flattened
+    //along mode 0. Not required for finite rank decomposition
+  //Max_R is the max rank, used by finite error computation
+  //skip allows rank R approximation to be built from skip
+    //many columns at a time
+  //ep_ALS is the threshold for the ALS minimization
+  //epsilon is the threshold for the CP decomposition
+
+  void compute(const int rank, const bool direct = true,
+               const bool r_test = false, const double max_R = 1e5,
+               const int skip = 1, const double ep_ALS = 0.1) {
+    Build(rank, direct, max_R, r_test, skip, ep_ALS);
+  }
+
+  void compute(const double epsilon = 1e-2, const bool direct = true,
+               const double max_R = 1e5, const int skip = 1,
+               const double ep_ALS = 0.1) {
+    int rank = (A.empty()) ? 0 : A[0].extent(0);
+    while (rank_test > epsilon && rank < max_R) {
+      rank += skip;
+      compute(rank, direct, true, max_R, skip, ep_ALS);
+      std::cout << "With rank " << rank << " there is a difference of "
+                << rank_test << std::endl;
+    }
+    std::cout << "The decomposition finished with rank " << rank
+              << " and epsilon " << rank_test << std::endl;
+  }
+  //computes ||T(0) - U(0)||F and returns the result
+  double difference() {
+    std::cout.precision(16);
+    auto rank = A[0].extent(1);
+    double diff = 0;
+    Tensor oldmat(tensor_ref.extent(0), size / tensor_ref.extent(0));
+    for (int i = 0; i < rank; i++) {
+      scal(A[0].extent(0), A[dim](i), std::begin(A[0]) + i, rank);
+    }
+    gemm(CblasNoTrans, CblasTrans, 1.0, A[0], generate_KRP(0, rank, true),
+         0.0, oldmat);
+    for (int i = 0; i < rank; i++) {
+      scal(A[0].extent(0), 1 / A[dim](i), std::begin(A[0]) + i, rank);
+    }
+    diff = std::abs(norm(Flatten(0) - oldmat));
+    return diff;
+  }
+  //Vector of (number of modes) factor matrices
+  //plus an array of weights, stored last.
+  std::vector<Tensor> A;  
+
+private:
+  ///////Global variables//////////
+
+  Tensor &tensor_ref; // this is a reference.
+  const int dim;
```

Comment:
better name: `ndim` (distinguish **number** of dimensions from a particular dimension index)